### PR TITLE
Fix crm_resource -a/-A output

### PR DESCRIPTION
**Pull request by clumens**
I'm not really sure of this at all, but hopefully it's a good direction to start in.  I have only worked on the text output.  XML is likely broken at best, segfaulty at worst.  I'll fix that should the text output changes be correct.  I could definitely use more CIB examples that I could use as test cases.

### CODE REVIEW

**Review by kgaillot — COMMENTED** [review]
A quick first look before I head out for the day, haven't had a chance to look at the actual output yet

File: lib/pacemaker/pcmk_output.c [hunk 1]
```diff
@@ -276,14 +276,21 @@ static int locations_list_xml(pcmk__output_t *out, va_list args) {
     return pcmk_rc_ok;
 }
 
+static void clear_allocating_flag(pe_working_set_t *data_set)
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/pacemaker/pcmk_output.c:319]: The colocations_header patch also needed a newline.  Fixed both.

File: lib/pacemaker/pcmk_output.c [hunk 2]
```diff
@@ -543,6 +574,7 @@ crmadmin_node_xml(pcmk__output_t *out, va_list args)
 static pcmk__message_entry_t fmt_functions[] = {
     { "colocations-list", "default", colocations_list },
     { "colocations-list", "xml", colocations_list_xml },
+    { "dependent-colocations-list", "default", dependent_colocations_list },
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/pacemaker/pcmk_output.c:572]: I haven't looked at the XML side at all yet, so it's possible it will require this same kind of split.

File: lib/pacemaker/pcmk_output.c [hunk 3]
```diff
@@ -101,17 +119,10 @@ static int colocations_list(pcmk__output_t *out, va_list args) {
...
-        score = score2char(cons->score);
-        if (cons->role_rh > RSC_ROLE_STARTED) {
-            out->list_item(out, NULL, "%s (score=%s, %s role=%s, id=%s",
-                           peer->id, score, dependents ? "needs" : "with",
-                           role2text(cons->role_rh), cons->id);
-        } else {
-            out->list_item(out, NULL, "%s (score=%s, id=%s",
-                           peer->id, score, cons->id);
-        }
+        hdr = colocations_header(peer, cons, TRUE);
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/pacemaker/pcmk_output.c]: Yeah, that's reasonable.  It's always nice to make this easier to follow.

File: include/crm/pengine/internal.h [hunk 4]
```diff
@@ -529,6 +529,7 @@ void pe_action_set_flag_reason(const char *function, long line, pe_action_t *act
 
 void pe__set_resource_flags_recursive(pe_resource_t *rsc, uint64_t flags);
 void pe__clear_resource_flags_recursive(pe_resource_t *rsc, uint64_t flags);
+void pe__clear_flag(pe_working_set_t *data_set, uint64_t flag);
```
Comment (kgaillot) [inline include/crm/pengine/internal.h:532]: Data sets have multiple flag types (resource flags, find flags, data set flags, node flags, graph flags, action flags, ordering flags, ...) so let's indicate somehow that this is for resource flags for all resources in the data set.

**Comment by kgaillot** [comment]
Just to pick a random example:

    CIB_file=cts/scheduler/whitebox-asymmetric.xml crm_resource -r nfs_mount -a
    webserver (score=INFINITY, id=colocation-webserver-nfs_mount-INFINITY):
      * Colocated with nfs_mount:
        * Locations:
          * Node 18builder (score=-INFINITY, id=location-nfs_mount-18builder--INFINITY)
          * Node 18node2 (score=INFINITY, id=location-nfs_mount-18node2-INFINITY)

What we want is something more like:

    * Locations:
      * Node 18builder (score=-INFINITY, id=location-nfs_mount-18builder--INFINITY)
      * Node 18node2 (score=INFINITY, id=location-nfs_mount-18node2-INFINITY)
    * Colocations with nfs_mount:
      * webserver (score=INFINITY, id=colocation-webserver-nfs_mount-INFINITY)

In other words, for each rsc_location constraint where rsc is $RSC_ID, show an item in the Locations list with the constraint's node, score, and id; and for each rsc_colocation constraint where with-rsc (aka ->rsc_rh->id) is $RSC_ID, shown an item in the Colocations list with the constraint's rsc, with-rsc, score, and id.

There's an argument to be made we should only show the constraint id's if verbose was specified. But that's optional.

For the recursive list, if webserver itself was used in rsc_location and/or rsc_colocation constraints, those would be shown beneath webserver.

File: lib/pacemaker/pcmk_output.c [hunk 5]
```diff
@@ -66,20 +66,15 @@ colocations_header(pe_resource_t *rsc, rsc_colocation_t *cons,
...
 
-PCMK__OUTPUT_ARGS("colocations-list", "pe_resource_t *", "gboolean", "gboolean")
+PCMK__OUTPUT_ARGS("colocations-list", "pe_resource_t *", "gboolean")
 static int colocations_list(pcmk__output_t *out, va_list args) {
     pe_resource_t *rsc = va_arg(args, pe_resource_t *);
-    gboolean dependents = va_arg(args, gboolean);
     gboolean recursive = va_arg(args, gboolean);
 
     GList *lpc = NULL;
     GList *list = rsc->rsc_cons;
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_output.c:75]: Since we're splitting the functions, we don't need the list variables anymore

Comment (clumens) [reply lib/pacemaker/pcmk_output.c:75]: Fixed, will show up in the next push (whenever that is).

**Comment by clumens** [comment]
This is still very much a WIP, but I've changed enough to where it's worth putting it out there for more people to look at.  I still have to address the XML side of things, and I haven't tested the recursive case yet, and it's entirely possible I've got the two new functions in 614c353 backwards.

**Comment by clumens** [comment]
Also, I'm getting sick of the
```
if (!printed_header) {
   ...
   printed_header = TRUE;
}
```
construct.  There's PCMK__OUTPUT_LIST_HEADER which almost addresses this, but not quite.  Maybe it's time to add something there (in a different PR).

File: include/crm/pengine/internal.h [hunk 6]
```diff
@@ -529,6 +529,7 @@ void pe_action_set_flag_reason(const char *function, long line, pe_action_t *act
 
 void pe__set_resource_flags_recursive(pe_resource_t *rsc, uint64_t flags);
 void pe__clear_resource_flags_recursive(pe_resource_t *rsc, uint64_t flags);
+void pe__clear_resource_flag_on_all(pe_working_set_t *data_set, uint64_t flag);
```
Comment (kgaillot) [inline include/crm/pengine/internal.h:532]: Nitpick, this should be "flags" because it will actually work with more than one flag or'ed together

File: lib/pacemaker/pcmk_output.c [hunk 7]
```diff
@@ -66,68 +66,46 @@ colocations_header(pe_resource_t *rsc, rsc_colocation_t *cons,
     return retval;
 }
 
-PCMK__OUTPUT_ARGS("colocations-list", "pe_resource_t *", "gboolean", "gboolean")
-static int colocations_list(pcmk__output_t *out, va_list args) {
+PCMK__OUTPUT_ARGS("rscs-colocated-with-list", "pe_resource_t *", "gboolean")
+static int rscs_colocated_with_list(pcmk__output_t *out, va_list args) {
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_output.c:70]: Yep, it's the reverse :)

for "A colocated with B", the constraint is added to A->rsc_cons and B->rsc_cons_lhs

So looping through rsc->rsc_cons shows all constraints where rsc is colocated with some other resource, and looping through rsc->rsc_cons_lhs shows all constraints where some other resource is colocated with rsc.

File: lib/pacemaker/pcmk_output.c [hunk 8]
```diff
@@ -227,6 +205,55 @@ static int colocations_list_xml(pcmk__output_t *out, va_list args) {
     return pcmk_rc_ok;
 }
 
+PCMK__OUTPUT_ARGS("rsc-is-colocated-with-list", "pe_resource_t *", "gboolean")
+static int rsc_is_colocated_with_list(pcmk__output_t *out, va_list args) {
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_output.c:207]: newline after int

File: tools/crm_resource.c [hunk 9]
```diff
@@ -379,10 +379,6 @@ static GOptionEntry query_entries[] = {
     { "locate", 'W', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, flag_cb,
       "Show node(s) currently running resource",
       NULL },
-    { "stack", 'A', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, flag_cb,
-      "Display the (co)location constraints that apply to a resource\n"
-      INDENT "and the resources is it colocated with",
-      NULL },
     { "constraints", 'a', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, flag_cb,
       "Display the (co)location constraints that apply to a resource",
```
Comment (kgaillot) [inline tools/crm_resource.c:383]: Also add "Optional: --recursive" and update the --recursive help to describe what it does in that case

**Review by kgaillot — COMMENTED** [review]
I think we're on the right track now :)

File: include/crm/pengine/internal.h (hunk 6 above)
Comment (clumens) [reply include/crm/pengine/internal.h:532]: Would this function be more broadly useful if instead of a `pe_working_set_t *`, it took a `GList *` of resources?

File: lib/pacemaker/pcmk_output.c (hunk 7 above)
Comment (clumens) [reply lib/pacemaker/pcmk_output.c:70]: Despite all the staring at those functions that I did, I still got them backwards.  I think they're right now.

File: lib/pacemaker/pcmk_output.c [hunk 10]
```diff
@@ -66,68 +66,46 @@ colocations_header(pe_resource_t *rsc, rsc_colocation_t *cons,
...
-
     if (pcmk_is_set(rsc->flags, pe_rsc_allocating)) {
         return pcmk_rc_no_output;
     }
 
     pe__set_resource_flags(rsc, pe_rsc_allocating);
-    for (lpc = list; lpc != NULL; lpc = lpc->next) {
+    for (GList *lpc = rsc->rsc_cons; lpc != NULL; lpc = lpc->next) {
         rsc_colocation_t *cons = (rsc_colocation_t *) lpc->data;
         pe_resource_t *peer = cons->rsc_rh;
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_output.c:83]: We don't need peer anymore either, after splitting the function in two

File: include/crm/pengine/internal.h (hunk 6 above)
Comment (kgaillot) [reply include/crm/pengine/internal.h:532]: I don't see any other such uses currently. It would be fine either way.

I noticed some more worms in the can though. Seeing the recursive function above made me wonder why -a/-A doesn't use something like that. The recursive function clears the flag(s) in the resource and all its children, and so on.

The original code does:

1. Clear pe_rsc_allocating on all top-level resources (so we can detect constraint loops, e.g. A colocated with B, B colocated with C, C colocated with A).

2. If the specified resource is marked as pe_rsc_allocating (of course it isn't at the beginning), skip it, otherwise mark it as pe_rsc_allocating. (Previously we would printf "loop" if it was marked -- I don't know when that disappeared.)

3. Loop through all colocation constraints involving the specified resource as the "with"/"right-hand" resource.
3a. If the other resource in the constraint has pe_rsc_allocating, we know we have a loop; skip this constraint.
3b. Otherwise, for -A, repeat from step 2 for the other resource. (***We introduced a bug, we specify rsc instead of peer***)
3c. Output a list item for the constraint.
3d. Print all location constraints for the *other* resource.

4. Print the specified resource and all its location constraints.

5. Clear pe_rsc_allocating on all top-level resources again. (Pre-existing bug: if any of the previously output constraints involved a child resource, it would not have its flag cleared. So I do think we should use pe__clear_resource_flags_recursive() in our new convenience function.)

6. If the specified resource is marked as pe_rsc_allocating (of course it isn't at the beginning), skip it, otherwise mark it as pe_rsc_allocating. (Previously we would printf "loop" if it was marked -- I don't know when that disappeared.)

7. Loop through all colocation constraints involving the specified resource as the "left-hand" resource.
7a. If the other resource in the constraint has pe_rsc_allocating, we know we have a loop; output "loop" and skip this constraint.
7b. Otherwise, output a list item for the constraint.
7c. Print all location constraints for the *other* resource.
7d. For -A, repeat from step 6 for the other resource. (***We introduced a bug, we specify rsc instead of peer***)

Obviously we can fix the bugs, but a bigger question is how the heck this information could be displayed in an intelligible way.

Comment (kgaillot) [reply include/crm/pengine/internal.h:532]: For test cases, we can create a single CIB with:

* Nodes N1, N2 (optionally more)
* Resources R1-R10
* Group G with resources GR1-GR3
* Clone C with primitive P
* No constraints for R1
* Colocation chain R2 with R3, R3 with R4, R4 with R5
* Colocation R10 with R4
* Location constraints for R2, R4, and R6 on a node or two each
* Colocation R7 with G
* Colocation R8 with GR2
* Colocation R9 with C

then run crm_resource -a with and without --recursive for each resource R1-R9, G, and C. We could do even more (cloned group, bundle, ...) but I think that would be sufficient.

File: lib/pacemaker/pcmk_output.c (hunk 10 above)
Comment (clumens) [reply lib/pacemaker/pcmk_output.c:83]: Removed.

File: include/crm/pengine/internal.h (hunk 6 above)
Comment (clumens) [reply include/crm/pengine/internal.h:532]: Quickly responding to two easy things in this comment...

> I don't see any other such uses currently. It would be fine either way.

I was just thinking in the future, it could be useful to be able to clear flags on any arbitrary list of resources.  I don't know if anything like that would ever come up, and it's a private function so we can make this change later should we need to, but I could also do it right now.

>     2. If the specified resource is marked as pe_rsc_allocating (of course it isn't at the beginning), skip it, otherwise mark it as pe_rsc_allocating. (Previously we would printf "loop" if it was marked -- I don't know when that disappeared.)

I removed this in a74e1193e005994ec4cfe1991e13bb61fff64de9, which is the commit that did the initial formatted output here.  I could put something back in its place.

Comment (kgaillot) [reply include/crm/pengine/internal.h:532]: I imagine something like:

    # crm_resource -a --recursive -r R4
    * Location preferences for R4:
       * Node N1 (score=100, id=I)
       * Node N2 (score=200, id=I)
    * Resources colocated with R4:
       * R3 with R4 (score=N, id=I)
          (here we would output "Location preferences for R3" if there were any)
          * Resources colocated with R3:
             * R2 with R3 (score=N, id=I)
          (here we would somehow output "Resources R3 is colocated with" excluding R4 if there were any) 
       * R10 with R4 (score=N, id=I)
         (here we would output location/colocations for R10 if there were any)
    * Resources R4 is colocated with:
       * R4 with R5 (score=N, id=I)
         (here we would output location/colocations for R5 if there were any)

or maybe more concisely:

    # crm_resource -a --recursive -r R4
    * Location preference for R4: score=100 on N1 (id=I)
    * Location preference for R4: score=200 on N2 (id=I)
    * R3 is colocated with R4 (score=N, id=I)
          * R2 is colocated with R3 (score=N, id=I)
    * R10 is colocated with R4 (score=N, id=I)
    * R4 is colocated with R5 (score=N, id=I)

One challenge I see is that a constraint listing might be a sublist heading or it might just be a list item.

(Also, I missed adding a colocation loop to the test CIB.)

Comment (kgaillot) [reply include/crm/pengine/internal.h:532]: Actually I see now that the original code is recursive only in the direction of the colocation, i.e. if R3 is colocated with R4 and we recursively show R4, we show constraints where some resource is colocated with R3, but not constraints where R3 is colocated with some resource other than R4. I suppose we can continue that approach to avoid having to filter the specified resource from the opposite constraints.

Comment (kgaillot) [reply include/crm/pengine/internal.h:532]: The idea being we're only showing constraints that involve the specified resource, directly or (if recursive) indirectly. That's why they showed it as a cascade, e.g.

    # crm_resource -a --recursive -r R4
    * Location preferences ...
    * Colocations:
       * R2 with R3
          * R3 with R4
             * R4 with R5
       * R10 with R4

Not sure what is the least confusing approach.

Comment (kgaillot) [reply include/crm/pengine/internal.h:532]: Also I'm second-guessing whether we should use --recursive. That currently means "and its children", but we'd be using it as "directly or indirectly". Maybe keep -a/-A and improve the help.

Comment (clumens) [reply include/crm/pengine/internal.h:532]: Dropped the patch to use --recursive.

**Comment by clumens** [comment]
I haven't addressed the most substantial comments about what this output should really look like, but I've made a bunch of other changes.

* Added test cases so you can see the shambles we are starting from.  I thought about adding XML tests as well, but it would double the already large number.  I might still do it.
* Dropped the --recursive patch, though I need to update docs still.
* Use PCMK__OUTPUT_LIST_HEADER and _FOOTER where appropriate, so there's a good bit less code.
* Renamed pcmk__clear_resource_flags_on_all again.

**Comment by clumens** [comment]
Okay, check out the test case output changes and code changes in the latest WIP commit.  I think that gets us a lot closer to coherent output, though I imagine it could still use work.  I'm getting happier with it (as well as getting a better understanding of how this is supposed to work).

**Comment by clumens** [comment]
There, added XML output too.  That has much more work to do, but I'll put it off until we're happy with the text output.  Then I only end up redoing one thing.

File: lib/pacemaker/pcmk_output.c [hunk 11]
```diff
@@ -145,15 +145,15 @@ rsc_is_colocated_with_list_xml(pcmk__output_t *out, va_list args) {
         rsc_colocation_t *cons = (rsc_colocation_t *) lpc->data;
 
         if (pcmk_is_set(cons->rsc_rh->flags, pe_rsc_allocating)) {
-            PCMK__OUTPUT_LIST_HEADER(out, FALSE, rc, "Colocations");
+            PCMK__OUTPUT_LIST_HEADER(out, FALSE, rc, "rsc-is-colocated-with");
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_output.c:154]: In XML all the colocations could be together in a single list, since rsc/with-rsc will indicate the direction of the constraint. But for practical reasons if it's easier to separate the two lists, that's fine too

**Review by kgaillot — COMMENTED** [review]
Man, I wish we had good names for the two directions of a colocation constraint. My brain hurts.

File: cts/cli/regression.crm_resource.exp [hunk 12]
```diff
@@ -0,0 +1,569 @@
+=#=#=#= Begin test: Check locations and constraints for prim1 =#=#=#=
+=#=#=#= End test: Check locations and constraints for prim1 - OK (0) =#=#=#=
+* Passed: crm_resource   - Check locations and constraints for prim1
+=#=#=#= Begin test: Recursively check locations and constraints for prim1 =#=#=#=
+=#=#=#= End test: Recursively check locations and constraints for prim1 - OK (0) =#=#=#=
+* Passed: crm_resource   - Recursively check locations and constraints for prim1
+=#=#=#= Begin test: Check locations and constraints for prim1 in XML =#=#=#=
+<pacemaker-result api-version="X" request="/home/clumens/src/pacemaker/tools/.libs/crm_resource -a -r prim1 --output-as=xml">
```
Comment (kgaillot) [inline cts/cli/regression.crm_resource.exp:8]: I don't think "request" will be the same on CI :)

File: cts/cli/regression.crm_resource.exp [hunk 13]
```diff
@@ -0,0 +1,569 @@
...
+=#=#=#= End test: Check locations and constraints for prim2 in XML - OK (0) =#=#=#=
+* Passed: crm_resource   - Check locations and constraints for prim2 in XML
+=#=#=#= Begin test: Recursively check locations and constraints for prim2 in XML =#=#=#=
+<pacemaker-result api-version="X" request="/home/clumens/src/pacemaker/tools/.libs/crm_resource -A -r prim2 --output-as=xml">
+  <locations>
+    <location host="cluster01" id="prim2-on-cluster1" score="INFINITY"/>
+  </locations>
+  <list name="rsc-is-colocated-with" count="1">
+    <colocation peer="prim3" id="colocation-prim2-prim3-INFINITY" score="INFINITY"/>
+  </list>
```
Comment (kgaillot) [inline cts/cli/regression.crm_resource.exp:60]: Recursive doesn't seem to be happening in XML

File: cts/cts-cli.in [hunk 14]
```diff
@@ -18,7 +18,7 @@ USAGE_TEXT="Usage: cts-cli [<options>]
 Options:
  --help          Display this text, then exit
  -V, --verbose   Display any differences from expected output
- -t 'TEST [...]' Run only specified tests (default: 'dates tools crm_mon acls validity upgrade rules')
+ -t 'TEST [...]' Run only specified tests (default: 'dates tools crm_mon crm_resource acls validity upgrade rules')
```
[the remark this answers is not in the document]
Comment (clumens) [reply cts/cts-cli.in:21]: I reorganized things a little bit so it's a much smaller block of code now.  I'll just include it with tools.

File: cts/cli/regression.crm_resource.exp (hunk 12 above)
Comment (clumens) [reply cts/cli/regression.crm_resource.exp:8]: Heh, whoops.  Our regex in cts/cts-cli.in needs to be updated to include crm_resource.  Fixed.

File: cts/cli/regression.crm_resource.exp (hunk 13 above)
Comment (clumens) [reply cts/cli/regression.crm_resource.exp:60]: No, it's not yet.  I was hoping to nail down the text side of things and then replicate the structure in the XML side.

File: lib/pacemaker/pcmk_output.c [hunk 15]
```diff
@@ -91,147 +91,153 @@ colocations_xml_node(pcmk__output_t *out, pe_resource_t *rsc,
...
+    int rc = pcmk_rc_no_output;
 
     if (pcmk_is_set(rsc->flags, pe_rsc_allocating)) {
-        return pcmk_rc_no_output;
+        return rc;
     }
 
     pe__set_resource_flags(rsc, pe_rsc_allocating);
-    for (lpc = list; lpc != NULL; lpc = lpc->next) {
+    for (GList *lpc = rsc->rsc_cons; lpc != NULL; lpc = lpc->next) {
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/pacemaker/pcmk_output.c:107]: I think I've taken care of this, or at least made it a lot easier to spot the problem in the future.  These terms also make my head hurt.

File: cts/cli/regression.crm_resource.exp (hunk 13 above)
Comment (clumens) [reply cts/cli/regression.crm_resource.exp:60]: Bah, it's easy enough to fix.  Fixed, and I can always look at the other problems later.

File: lib/pacemaker/pcmk_output.c (hunk 11 above)
Comment (clumens) [reply lib/pacemaker/pcmk_output.c:154]: It might be reasonable to just flatten the whole structure for XML, putting all the recursive results into a single flat list.

**Comment by clumens** [comment]
Differences from the last push:

* I split the last big WIP/"fix the output" patch into several smaller ones that should hopefully make it easier to follow along.  One of these just fixes the bugs in the recursive output, for instance.
* XML output functions should now be recursive.
* The XML output is flattened into a single list, with location constraints nested under the proper tag.
* The patch that splits colocations_list apart has been turned into two patches, one of which does just the split, and the other of which makes further changes.

I still need to update the XML schema, but I'm saving that for when we've got the XML output completely figured out.

File: lib/pacemaker/pcmk_output.c [hunk 16]
```diff
@@ -150,10 +150,10 @@ rsc_is_colocated_with_list_xml(pcmk__output_t *out, va_list args) {
 
         PCMK__OUTPUT_LIST_HEADER(out, FALSE, rc, "rsc-is-colocated-with");
         colocations_xml_node(out, cons->rsc_rh, cons, FALSE);
-        out->message(out, "locations-list", cons->rsc_rh);
+        out->message(out, "locations-list", cons->rsc_lh);
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_output.c:153]: This should still be rh

File: lib/pacemaker/pcmk_output.c [hunk 17]
```diff
@@ -175,12 +175,9 @@ rscs_colocated_with_list(pcmk__output_t *out, va_list args) {
...
 
         if (pcmk_is_set(cons->rsc_lh->flags, pe_rsc_allocating)) {
-            continue;
-        }
-
-        if (recursive) {
             PCMK__OUTPUT_LIST_HEADER(out, FALSE, rc, "Resources colocated with %s", rsc->id);
-            out->message(out, "rscs-colocated-with-list", cons->rsc_lh, recursive);
+            out->list_item(out, NULL, "%s (id=%s - loop)", cons->rsc_lh->id, cons->id);
+            continue;
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_output.c:184]: Good catch. BTW we could move the list header before the is-set check since it's identical below now

File: cts/cli/regression.tools.exp [hunk 18]
```diff
@@ -3471,58 +3470,56 @@ Removing constraint: cli-prefer-dummy
...
+    <rsc_colocation id="colocation-prim2-prim3-INFINITY" rsc="prim2" with-rsc="prim3" score="INFINITY" node-attribute="" rsc-role="Unknown" with-rsc-role="Unknown">
+      <locations>
         <location host="cluster01" id="prim2-on-cluster1" score="INFINITY"/>
-      </list>
-    </resource>
-    <colocations>
-      <rsc_colocation id="colocation-prim2-prim3-INFINITY" rsc="prim2" with-rsc="prim3" score="INFINITY" node-attribute="" rsc-role="Unknown" with-rsc-role="Unknown"/>
-    </colocations>
+      </locations>
+    </rsc_colocation>
```
Comment (kgaillot) [inline cts/cli/regression.tools.exp:3522]: The CIB uses rsc_location (aka XML_CONS_TAG_RSC_LOCATION) for location constraints (and "node" instead of "host"), so we might as well continue with that. Also, we can go even flatter and ditch the locations/colocations elements, rsc_location and rsc_colocation will give enough info.

**Review by kgaillot — COMMENTED** [review]
Looking great, I think the average user might actually be able to make sense of -a/-A by the time we're through :)

File: tools/crm_resource.c [hunk 19]
```diff
@@ -1591,7 +1591,8 @@ main(int argc, char **argv)
          */
         if (options.rsc_cmd == cmd_list_resources || options.rsc_cmd == cmd_query_xml ||
             options.rsc_cmd == cmd_query_raw_xml || options.rsc_cmd == cmd_list_active_ops ||
-            options.rsc_cmd == cmd_list_all_ops) {
+            options.rsc_cmd == cmd_list_all_ops || options.rsc_cmd == cmd_colocations ||
+            options.rsc_cmd == cmd_colocations_deep) {
```
[the remark this answers is not in the document]
Comment (clumens) [reply tools/crm_resource.c:1595]: Yeah, agreed.  Fixed.

File: lib/pacemaker/pcmk_output.c (hunk 17 above)
Comment (clumens) [reply lib/pacemaker/pcmk_output.c:184]: Fixed.

File: cts/cli/regression.tools.exp (hunk 18 above)
Comment (clumens) [reply cts/cli/regression.tools.exp:3522]: Are you thinking something like this instead:
```
<constraints>
  <location node="cluster01" id="prim2-on-cluster1" score="INFINITY"/>
  <rsc_location id="colocation-prim2-prim3-INFINITY" rsc="prim2" with-rsc="prim3" score="INFINITY" node-attribute="" rsc-role="Unknown" with-rsc-role="Unknown">
    <location node="cluster01" id="prim2-on-cluster1" score="INFINITY"/>
  </rsc_location>
</constraints>
```

Comment (kgaillot) [reply cts/cli/regression.tools.exp:3522]: Like:
```
<constraints>
  <rsc_location node="cluster01" id="prim2-on-cluster1" score="INFINITY"/>
  <rsc_colocation id="colocation-prim2-prim3-INFINITY" rsc="prim2" with-rsc="prim3" score="INFINITY" node-attribute="" rsc-role="Unknown" with-rsc-role="Unknown"/>
  <rsc_location node="cluster01" id="prim2-on-cluster1" score="INFINITY"/>
</constraints>
```

Comment (kgaillot) [reply cts/cli/regression.tools.exp:3522]: Edited XML in previous comment. Looking at it again, I'm not sure why the location constraint is doubled in this PR's output. Maybe the recursive location message is being given the original resource rather than the peer. I also didn't notice that rsc was missing from the location constraints.

Comment (clumens) [reply cts/cli/regression.tools.exp:3522]: I can fix all those in a followup commit in this PR.  The rsc has always been missing from location constraints because it wasn't necessary - we had that tree structure to the output, so you could figure out the resource from context.  Once I started messing around with it, that was no longer the case.  So yeah, I'll add it in.

Comment (clumens) [reply cts/cli/regression.tools.exp:3522]: Location constraints are doubled only for the resource given on the command line with "-r <rsc>", I suppose because of the extra locations-list call in stacks_and_constraints.

Comment (clumens) [reply cts/cli/regression.tools.exp:3522]: And appears to be fixed by replacing an errant cons->rsc_lh that should have been an rh.

File: lib/pacemaker/pcmk_output.c [hunk 20]
```diff
@@ -66,6 +66,24 @@ colocations_header(pe_resource_t *rsc, rsc_colocation_t *cons,
...
+static void
+colocations_xml_node(pcmk__output_t *out, pe_resource_t *rsc,
+                     rsc_colocation_t *cons, gboolean dependents) {
+    char *score = NULL;
+
+    score = score2char(cons->score);
+    pcmk__output_create_xml_node(out, "rsc_colocation",
+                                 "id", cons->id,
+                                 "rsc", dependents ? cons->rsc_rh->id : cons->rsc_lh->id,
+                                 "with-rsc", dependents ? cons->rsc_lh->id : cons->rsc_rh->id,
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/pacemaker/pcmk_output.c]: Making this change results in output like this:
```
<pacemaker-result api-version="2.3" request="/home/clumens/src/pacemaker/tools/.libs/crm_resource -A -r prim13 --output-as=xml">
  <constraints>
    <rsc_colocation id="colocation-prim13-prim11-INFINITY" rsc="prim13" with-rsc="prim11" score="INFINITY" node-attribute="" rsc-role="Unknown" with-rsc-role="Unknown"/>
    <rsc_colocation id="colocation-prim11-prim12-INFINITY" rsc="prim11" with-rsc="prim12" score="INFINITY" node-attribute="" rsc-role="Unknown" with-rsc-role="Unknown"/>
    <rsc_colocation id="colocation-prim12-prim13-INFINITY" rsc="prim12" with-rsc="prim13" score="INFINITY" node-attribute="" rsc-role="Unknown" with-rsc-role="Unknown"/>
    <rsc_colocation id="colocation-prim13-prim11-INFINITY" rsc="prim13" with-rsc="prim11" score="INFINITY" node-attribute="" rsc-role="Unknown" with-rsc-role="Unknown"/>
    <rsc_colocation id="colocation-prim11-prim12-INFINITY" rsc="prim11" with-rsc="prim12" score="INFINITY" node-attribute="" rsc-role="Unknown" with-rsc-role="Unknown"/>
    <rsc_colocation id="colocation-prim12-prim13-INFINITY" rsc="prim12" with-rsc="prim13" score="INFINITY" node-attribute="" rsc-role="Unknown" with-rsc-role="Unknown"/>
  </constraints>
  <status code="0" message="OK"/>
</pacemaker-result>
```

As you can see, everything is duplicated.  Commenting out one of the out->message blocks in stacks_and_constraints_xml removes the duplicate.  This makes me think one of these functions still isn't doing the right thing when it comes to lh vs. rh.

File: lib/pacemaker/pcmk_output.c [hunk 21]
```diff
@@ -84,69 +84,145 @@ colocations_xml_node(pcmk__output_t *out, pe_resource_t *rsc,
     free(score);
 }
 
-PCMK__OUTPUT_ARGS("colocations-list", "pe_resource_t *", "gboolean", "gboolean")
-static int colocations_list(pcmk__output_t *out, va_list args) {
+PCMK__OUTPUT_ARGS("rsc-is-colocated-with-list", "pe_resource_t *", "gboolean")
+static int
+rsc_is_colocated_with_list(pcmk__output_t *out, va_list args) {
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/pacemaker/pcmk_output.c:89]: I think those are reasonable names.  I'd just need to remember how to map everything from the old names to the new.

**Comment by clumens** [comment]
I'm going to push this one more time, if for no other reason than to have a backup of all this hard work in case something happens over the break.

File: lib/pacemaker/pcmk_output.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_output.c]: That particular example may be an artifact of it being your loop case.

[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_output.c]: Either that or you're right and they're all loops :)

**Comment by kgaillot** [comment]
This is looking good to me. The prim11/12/13 test is the loop test, so it makes some sense that they're doubled (the relationships occur in both directions). I suppose we could skip all further constraints once we get any loop to avoid that, but I'm not sure it matters, and we should probably add an exit status for "loop" if we do that. Leaving it as-is is probably fine.

If you're reasonably happy with it, add the XML schema changes and I'll merge it. We can always do more corrections after the holidays.

**Comment by clumens** [comment]
Yeah I am happy with how this looks right now.  I'm sure there's some bug hiding in there, but it's vastly better than what we previously had.  I made some changes to 393342e to only add certain attributes if they are set, which then required re-running cts on a couple patches.  With that, all the CIB chunks in cts/cli/constraints.xml validate.

**Comment by kgaillot** [comment]
Awesome. BTW we ended up with api-result-2.5 for the 2.0.5 release, so we'll have to rename crm_resource-2.5 to 2.6.
